text/MichalStobierski/opis.tex

```diff
@@ -0,0 +1,129 @@
+\subsection{Dekompresja}
+W niniejszym tekście nie będziemy poświęcać wiele miejsca dekompresji, gdyż z algorytmicznego punktu widzenia jest ona taka podobna dla wszystkich algorytmów w klasie "prefiksowych". To, co tu być może warto podkreślić to fakt, że dekompresja odbywa się jedynie z pomocą $P_{d}$ (zdefiniowanego identycznie jak w kompresorze), nie ma zatem znaczenia w jaki sposób parsowaliśmy wyjście kompresora (wybór $P_{o}$).
+\newpage
+\section{Algorytm i dowód poprawności}
+W ramach wprowadzenia do zasadniczej części przekonajmy się, że rzeczywiście  $P_{o}$ realizowany w postaci algorytmu zachłannego może dać wyniki dalekie od oczekiwanych:
+
+\begin{theorem}[Nieoptymalność zachłannego parsera \cite{DBLP:conf/soda/MatiasS99}]
+Zachłanny parser wyjścia w każdej iteracji algorytmu wybiera najdalej sięgający prefiks nieskompresowanej części tekstu. Istnieją słowa, które używając słownika $D$ z alg. LZW można zakodować przy użyciu $O(l)$ fraz, a dla których zachłanny $P_{o}$ używa $O(l^{3/2})$ fraz.
+\end{theorem}
+\begin{proof}[Szkic dowodu]
+Niech $k$ będzie l. pierwszą. Zdefiniujmy $\Sigma=\{0,1,...,k+\sqrt{k}\}$, $R=1,...,k$, $R_{i} = R^{i}$, $S=1,2,1,2,3,...,1,2,...,k$. Za słowo do skompresowania weźmy \[T=0,S,k+1,0,R_{1},1,k+2,0,R_{2},1,...,k+\sqrt{k},0,R_{\sqrt{k-1}},1\]
+Pozostaje przeanalizować, co zrobi zachłanny parser, a co można uzyskać w sprytniejszy sposób (np. podział na $O(k)$ fraz).
+\end{proof}
+
+Teraz pora na najważniejszą część niniejszego rozważania. Okazuje się, że \textbf{jeżeli zamiast wybierania prefiksu, który w danej iteracji pokryje najwięcej tekstu wybierzemy taki, który spowoduje największy postęp w następnej iteracji, to uzyskamy algorytm optymalny (oznaczmy go $FP$)}. Stwierdzenie to formalizuje poniższe twierdzenie:
+
+\begin{theorem}[Optymalność $FP$ \cite{DBLP:conf/soda/MatiasS99}]
+Dla jakiegokolwiek $P_{d}$ które tworzy słownik $D$ mający własność prefiksowości przez cały czas trwania algorytmu, $FP$ generuje skompresowany ciąg o najmniejszej liczbie użytych kodów (fraz), bez względu na dobór tekstu $T$.
+\end{theorem}
+\begin{proof}[Dowód]
+Niech $P_{d}$ będzie takie jak w wypowiedzi. Niech $C$ i $C'$ to będą kompresory używające kolejno $P_{o} = FP$ i jakiegokolwiek inny sposób parsowania wyjścia. Chcemy pokazać, że \begin{center}
+$\#output(C) \leq \#output(C')$
+\end{center}
+
+Niech $i$-ta fraza wypisana przez $C$ kończy się na pozycji $E(i)$. 
+Niech najdalej sięgająca fraza w $D$ będąca prefiksem $T[E(i)+1:n]$ kończy się na pozycji $L(i)$.
+Analogicznie definiujemy $E'(i)$ oraz $L'(i)$ dla $C$.
+Jeżeli pokażemy, że $\forall_{i}$ $L(i) \geq L'(i)$ to otrzymujemy tezę. Dowód przebiega indukcyjnie:\\
+Baza: jest trywialnie spełniona, bo $C$ i $C'$ startują z tej samej pozycji.\\
+Krok indukcyjny: zauważmy, że $t_{0} = T[E'(i)+1:L'(i)]$ jest frazą w $D$ (z definicji $L'$) i zachodzi: 
```

Comment:
Maybe I'm missing something, but $D$ is a dictionary with respect to $C$, right? Because I'd expect that both dictionaries will be different, so i don't see how $t_0$ has to be in $D$ in virtue of $L'$.

Reply:
So apparently I didn't clarify this enough, but the whole idea is, having the same **dictionary creation** method

 _Dla jakiegokolwiek $P{d}$ które tworzy słownik $D$ mający własność prefiksowości przez cały czas trwania algorytmu (...)_,

our **flexible parsing** is better than any other method. So fixed $P_{d}$ implies same $D$ for both $C$ and $C'$.